lib/netext/dialer.go

```diff
@@ -149,6 +141,57 @@ func (d *Dialer) GetTrail(
 	}
 }
 
+func (d *Dialer) dialAddr(addr string) (string, error) {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return "", err
+	}
+
+	// lookup for domain defined in Hosts option before trying to resolve DNS.
+	remote, ok := d.Hosts[addr]
+	if !ok {
+		remote, err = d.fetchHostAddress(host, port)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	for _, ipnet := range d.Blacklist {
+		if (*net.IPNet)(ipnet).Contains(remote.IP) {
+			return "", BlackListedIPError{ip: remote.IP, net: ipnet}
+		}
+	}
+
+	if remote.Port == 0 && port != "" {
+		var err error
+		remote, err = lib.NewHostAddress(remote.IP, port)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return remote.String(), nil
+}
+
+func (d *Dialer) fetchHostAddress(host, port string) (*lib.HostAddress, error) {
+	remote := d.Hosts[host]
+	if remote != nil {
+		return remote, nil
+	}
+	ip, err := d.Resolver.FetchOne(host)
+	if err != nil {
+		return nil, err
+	}
+	if ip == nil {
+		ip = net.ParseIP(host)
+	}
+	if ip == nil {
+		return nil, errors.Errorf("lookup %s: no such host", host)
```

Comment:
You could try `ParseIP()` right after calling `SplitHostPort()`, and if it is an IP you can skip the DNS lookup, which would avoid returning this error masking as a resolver error. See what I mean in 4c0e6639.

Reply:
that's a great idea. I've pushed a new commit with that change.